### PR TITLE
Use hiding list to reduce lag from many annotations

### PR DESCRIPTION
**Pull request by kaikue**
Instead of keeping every annotation element in the list at once (causing significant lag when recalculating layout), use a special list that applies `display:none` to offscreen elements so their dimensions can be ignored in layout.

This fixes https://github.com/seung-lab/neuroglancer/issues/364.

Also, this will conflict with https://github.com/seung-lab/neuroglancer/pull/360, so when that is integrated some changes will be need to made (probably flattening the DOM from a tree to a flat list).

### CODE REVIEW

File: src/neuroglancer/ui/hiding_list.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    this.updateScrollHeight();
+    this.updateScrollAreaPos();
+  }
+
+  removeElement(element: HTMLElement) {
+    this.resizeObserver.unobserve(element);
+    this.unhideElement(element);
+    const h = element.offsetHeight;
+
+    const i = this.elementIndices.get(element)!;
```
Comment (manuel-castro) [inline src/neuroglancer/ui/hiding_list.ts]: I think elementIndex would be a more descriptive name

File: src/neuroglancer/ui/hiding_list.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    this.scrollbarFiller.style.height = (this.totalH + 10) + 'px';
+  }
+
+  private shiftHeightsAfter(startIndex: number, delta: number) {
+    for (let i = startIndex; i < this.elementHeights.length; i++) {
+      this.elementHeights[i][1] += delta;
+    }
+  }
+
+  private findHeight(h: number) {
```
Comment (manuel-castro) [inline src/neuroglancer/ui/hiding_list.ts]: A comment explaining why we need to find this specific height, and mentioning that this function performs a binary search, would be quite helpful. I also think the variable names could be a little more descriptive (l -> left, r -> right, m -> mid).

File: src/neuroglancer/ui/hiding_list.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
+import ResizeObserver from 'resize-observer-polyfill';
+
+export class HidingList {
```
Comment (manuel-castro) [inline src/neuroglancer/ui/hiding_list.ts]: Cool! I really like how you organized this class.

**Comment by kaikue** [comment]
Names have been updated to be more descriptive overall. I also changed some instances of "height" to "Y" to make the distinction between "Y position in the list" and "height of list element" more clear.

**Pull request merged**